src/Lib/package.cpp

```diff
@@ -34,17 +36,61 @@ const std::string Package::GetFullyQualifiedName(void) {
   return Package::GetName() + " " + Package::GetVersion();
 }
 
+const void Package::assignDataDir(void) {
+  const char *possible[] = {".", "..", DATADIR, NULL};
+  int i = 0;
+
+  while (possible[i] != NULL) {
+    char map[1024] = "";
+    size_t r = snprintf(map, sizeof map, "%s/maps/Two clans.npm", possible[i]);
+    if (r >= sizeof map) {
+      // handle error
+    }
+
+    std::ifstream file(map);
```

Comment:
OMG!! what is up with the `bool exists(const char* filename) { return PHYSFS_exists(filename); }` function that's being used throughout the code? And how are we even able to use it? I spent two hours trying to pass the test, and finally I could see it was segfaulting. I switched to this method from the std lib, and it worked immediately.

Reply:
Weird, What does valgrind say?

Reply:
When the map didn't exist (which was correct, because it was checking '.').

```
meson_build_root: /home/andy/src/netpanzer/netpanzer/_build                                                                 
cwd: /home/andy/src/netpanzer/netpanzer/_build
map: ./maps/Two clans.npm

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff733d3c0 in ?? () from /usr/lib/libphysfs.so.1
(gdb) info stack
#0  0x00007ffff733d3c0 in ?? () from /usr/lib/libphysfs.so.1
#1  0x00007ffff73404fd in PHYSFS_exists () from /usr/lib/libphysfs.so.1
#2  0x00007ffff7d1c5db in filesystem::exists (filename=0x7fffffffdcb0 "./maps/Two clans.npm")
    at ../src/Lib/Util/FileSystem.cpp:209
#3  0x00007ffff7ca8a27 in Package::assignDataDir () at ../src/Lib/package.cpp:51
#4  0x00005555555564f4 in main () at ../src/Lib/package.cpp:84
```

Reply:
Valgrind should give you more info.

Reply:
This was my code: https://github.com/netpanzer/netpanzer/compare/iss-71...iss-71-util-filesys-segfault

```
[andy@prometheus tests](iss-71)$ valgrind --leak-check=full ./test_package
==552601== Memcheck, a memory error detector
==552601== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==552601== Using Valgrind-3.22.0 and LibVEX; rerun with -h for copyright info
==552601== Command: ./test_package
==552601== 
==552601== Invalid read of size 8
==552601==    at 0x54BB3C0: ??? (in /usr/lib/libphysfs.so.3.2.0)
==552601==    by 0x54BE4FC: PHYSFS_exists (in /usr/lib/libphysfs.so.3.2.0)
==552601==    by 0x49EF55A: filesystem::exists(char const*) (FileSystem.cpp:209)
==552601==    by 0x497B9FE: Package::assignDataDir() (package.cpp:57)
==552601==    by 0x10A485: test_datadir() (package.cpp:90)
==552601==    by 0x10A58D: main (package.cpp:99)
==552601==  Address 0x28 is not stack'd, malloc'd or (recently) free'd
==552601== 
==552601== 
==552601== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==552601==  Access not within mapped region at address 0x28
==552601==    at 0x54BB3C0: ??? (in /usr/lib/libphysfs.so.3.2.0)
==552601==    by 0x54BE4FC: PHYSFS_exists (in /usr/lib/libphysfs.so.3.2.0)
==552601==    by 0x49EF55A: filesystem::exists(char const*) (FileSystem.cpp:209)
==552601==    by 0x497B9FE: Package::assignDataDir() (package.cpp:57)
==552601==    by 0x10A485: test_datadir() (package.cpp:90)
==552601==    by 0x10A58D: main (package.cpp:99)
==552601==  If you believe this happened as a result of a stack
==552601==  overflow in your program's main thread (unlikely but
==552601==  possible), you can try to increase the size of the
==552601==  main thread stack using the --main-stacksize= flag.
==552601==  The main thread stack size used in this run was 8388608.
==552601== 
==552601== HEAP SUMMARY:
==552601==     in use at exit: 93,234 bytes in 27 blocks
==552601==   total heap usage: 29 allocs, 2 frees, 93,620 bytes allocated
==552601== 
==552601== LEAK SUMMARY:
==552601==    definitely lost: 0 bytes in 0 blocks
==552601==    indirectly lost: 0 bytes in 0 blocks
==552601==      possibly lost: 0 bytes in 0 blocks
==552601==    still reachable: 93,234 bytes in 27 blocks
==552601==         suppressed: 0 bytes in 0 blocks
==552601== Reachable blocks (those to which a pointer was found) are not shown.
==552601== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==552601== 
==552601== For lists of detected and suppressed errors, rerun with: -s
==552601== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
Segmentation fault (core dumped)
```

This is the output that's in '_build/meson-logs/testlog.txt' when I build with the sanitizer:

```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==551845==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000028 (pc 0x7fe1f92483c0 bp 0x7ffecef1b480 sp 0x7ffecef1b400 T0)
==551845==The signal is caused by a READ memory access.
==551845==Hint: address points to the zero page.
    #0 0x7fe1f92483c0  (/usr/lib/libphysfs.so.1+0x83c0) (BuildId: 08e4702f577e7ce0e7dafc8c32392b736e888d8e)
    #1 0x7fe1f924b4fc in PHYSFS_exists (/usr/lib/libphysfs.so.1+0xb4fc) (BuildId: 08e4702f577e7ce0e7dafc8c32392b736e888d8e)
    #2 0x7fe1fb0be9b4 in filesystem::exists(char const*) ../src/Lib/Util/FileSystem.cpp:209
    #3 0x7fe1fae40511 in Package::assignDataDir() ../src/Lib/package.cpp:57
    #4 0x558e604b993e in test_datadir() ../src/Lib/package.cpp:90
    #5 0x558e604b9b62 in main ../src/Lib/package.cpp:99
    #6 0x7fe1f96edccf  (/usr/lib/libc.so.6+0x27ccf) (BuildId: 8bfe03f6bf9b6a6e2591babd0bbc266837d8f658)
    #7 0x7fe1f96edd89 in __libc_start_main (/usr/lib/libc.so.6+0x27d89) (BuildId: 8bfe03f6bf9b6a6e2591babd0bbc266837d8f658)
    #8 0x558e604b9354 in _start (/home/andy/src/netpanzer/netpanzer/_sanitize/tests/test_package+0x6354) (BuildId: 5df90c4295327a6c95f64ae0fec658df45fb1f5a)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV (/usr/lib/libphysfs.so.1+0x83c0) (BuildId: 08e4702f577e7ce0e7dafc8c32392b736e888d8e) 
==551845==ABORTING
```

Reply:
Good to know. I think we can just use the stdlib? Why do we need a library for that? Wonder what the thought process was.

Wonder if the library is failing on double-dot in the file path.